test/jdk/java/lang/Thread/crac/ThreadSleepOnCRPauseTest.java

```diff
@@ -0,0 +1,140 @@
+// CA 94089 USA or visit www.azul.com if you need additional information or
+// have any questions.
+
+/*
+ * @test ThreadSleepOnCRPauseTest.java
+ * @requires (os.family == "linux")
+ * @library /test/lib
+ * @summary check if the Thread.sleep() will be completed on restore immediately
+ *          if its end time fell on the CRaC pause period
+ *          (i.e. between the checkpoint and restore)
+ * @run main ThreadSleepOnCRPauseTest
+ */
+
+
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+import java.util.concurrent.CountDownLatch;
+
+
+public class ThreadSleepOnCRPauseTest {
+
+    private final static long SLEEP   = 1500; // [ms]
+    private final static long CRPAUSE = 2500; // [ms]
+
+    private final static long EPS = 300_000_000; // [ns], i.e. 0.3 s
+
+    private final CountDownLatch sleepLatch = new CountDownLatch(1);
+
+    private volatile long end1Time = 0;
+    private volatile long end2Time = 0;
+
+
+    private void runTest() throws Exception {
+
+        Runnable r1 = () -> {
+
+            try {
+                sleepLatch.await();
+                Thread.sleep(SLEEP);
+                end1Time = System.nanoTime();
+            } catch (InterruptedException ie) {
+                throw new RuntimeException(ie);
+            }
+        };
+
+        Runnable r2 = () -> {
+
+            try {
+                sleepLatch.await();
+                Thread.sleep(SLEEP, 20);
+                end2Time = System.nanoTime();
+            } catch (InterruptedException ie) {
+                throw new RuntimeException(ie);
+            }
+        };
+
+        Thread t1 = new Thread(r1), t2 = new Thread(r2);
+        t1.start();
+        t2.start();
+        sleepLatch.countDown(); // sleep simultaneously
+
+        while ((t1.getState() != Thread.State.TIMED_WAITING) &&
+               (t2.getState() != Thread.State.TIMED_WAITING)) {
+            Thread.onSpinWait();
+        }
+
+        long beforeCheckpoint = System.nanoTime();
+
+        jdk.crac.Core.checkpointRestore();
+
+        long afterRestore = System.nanoTime();
+
+        t1.join();
+        t2.join();
+
+        // being paranoid #1
+        long pause = afterRestore - beforeCheckpoint;
+        if (pause < 1_000_000 * CRPAUSE - EPS) {
+            throw new RuntimeException("the CR pause was less than " + CRPAUSE + " msec");
+        }
+
+        // being paranoid #2
+        if (end1Time < beforeCheckpoint + EPS || end2Time < beforeCheckpoint + EPS) {
+            throw new RuntimeException("sleep might probably was finished " +
+                "before the checkpoint for at least one thread");
+        }
+
+        if (Math.max(
+                Math.abs(afterRestore - end1Time),
+                Math.abs(afterRestore - end2Time))  > EPS) {
+            throw new RuntimeException("the sleeping threads were not " +
+                    "finished in " + EPS + " ns after the restore");
+        }
```

Comment:
It would be useful to print the difference for human eyes

Reply:
ok